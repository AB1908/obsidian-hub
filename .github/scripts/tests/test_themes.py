import approvaltests
from approvaltests import verify_as_json, verify
from approvaltests.storyboard import StoryBoard

import utils
from tests.test_templates import JINJA_TEMPLATES_DIR
from themes import get_theme_downloads, collect_data_for_theme, ThemeList
from utils import THEMES_JSON_FILE, get_json_from_github


def test_get_theme_downloads() -> None:
    downloads = get_theme_downloads()
    assert len(downloads) >= 117, len(downloads)

    key_of_first_download = list(downloads.keys())[0]
    download = downloads[key_of_first_download]
    assert download["id"] == key_of_first_download, download["id"]
    assert download["download"] > 0, download["download"]

    # TODO Scrub the download numbers
    # TODO Don't add new themes
    # verify_as_json(downloads)


def test_get_community_themes() -> None:
    theme_list: ThemeList = get_json_from_github(THEMES_JSON_FILE)
    assert len(theme_list) >= 115, len(theme_list)

    first_theme = theme_list[0]
    assert len(first_theme["modes"]) > 0
    assert first_theme["author"] != ""
    assert first_theme["name"] != ""
    assert first_theme["screenshot"] != ""

    # TODO Don't add new themes - maybe get a fixed revision
    # verify_as_json(theme_list)


def test_collect_data_for_theme_with_settings() -> None:
    theme_name = "Minimal"
    verify_theme_data(theme_name)


def test_collect_data_for_theme_without_settings() -> None:
    theme_name = "Christmas"
    verify_theme_data(theme_name)


def verify_theme_data(theme_name: str) -> None:
    s = StoryBoard()

    template = utils.get_template_from_directory(JINJA_TEMPLATES_DIR, "theme.md.jinja")
    theme_downloads = get_theme_downloads()

    theme_list: ThemeList = get_json_from_github(THEMES_JSON_FILE)
    theme = None
    for t in theme_list:
        if t["name"] == theme_name:
            theme = t
            break
    assert theme
    assert theme["name"] == theme_name
    assert len(theme["modes"]) > 0
    assert theme["author"] != ""

    s.add_frame(approvaltests.utils.to_json(theme))

    name = collect_data_for_theme(theme, theme_downloads, template)
    assert name == theme_name
    assert theme["user"] != ""
    s.add_frame(approvaltests.utils.to_json(theme))

    # TODO Scrub the download numbers
    # verify(s)
