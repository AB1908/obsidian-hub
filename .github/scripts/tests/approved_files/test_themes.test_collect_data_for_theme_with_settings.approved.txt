Initial:
{
    "author": "kepano",
    "modes": [
        "dark",
        "light"
    ],
    "name": "Minimal",
    "repo": "kepano/obsidian-minimal",
    "screenshot": "dark-simple.png"
}

Frame #1:
{
    "author": "kepano",
    "branch": "master",
    "download_count": 60237,
    "modes": "[[Dark-mode themes|dark]], [[Light-mode themes|light]]",
    "name": "Minimal",
    "plugins": {
        "community": [
            "buttons|Buttons",
            "dataview|Dataview",
            "calendar|Calendar",
            "obsidian-checklist-plugin|Checklist",
            "obsidian-codemirror-options|CodeMirror Options",
            "obsidian-dictionary-plugin|Dictionary",
            "obsidian-embedded-note-titles|Embedded Note Titles",
            "obsidian-excalidraw-plugin|Excalidraw",
            "obsidian-git|Obsidian Git",
            "obsidian-hider|Hider",
            "obsidian-kanban|Kanban",
            "obsidian-minimal-settings|Minimal Theme Settings",
            "obsidian-outliner|Outliner",
            "obsidian-system-dark-mode|System Dark Mode",
            "obsidian-style-settings|Style Settings",
            "quickadd|QuickAdd",
            "sliding-panes-obsidian|Sliding Panes (Andy Matuschak Mode)"
        ],
        "core": [
            "Obsidian Core Plugins#Backlinks|Backlinks",
            "Obsidian Core Plugins#Command palette|Command palette",
            "Obsidian Core Plugins#Daily notes|Daily notes",
            "Obsidian Core Plugins#File explorer|File explorer",
            "Obsidian Core Plugins#File recovery|File recovery",
            "Obsidian Core Plugins#Search|Search",
            "Obsidian Core Plugins#Graph view|Graph view",
            "Obsidian Core Plugins#Outgoing Links|Outgoing Links",
            "Obsidian Core Plugins#Outline|Outline",
            "Obsidian Core Plugins#Page preview|Page preview",
            "Obsidian Core Plugins#Publish|Publish",
            "Obsidian Core Plugins#Random note|Random note",
            "Obsidian Core Plugins#Starred notes|Starred notes",
            "Obsidian Core Plugins#Quick switcher|Quick switcher",
            "Obsidian Core Plugins#Sync|Sync",
            "Obsidian Core Plugins#Tag pane|Tag pane",
            "Obsidian Core Plugins#Word count|Word count"
        ]
    },
    "repo": "kepano/obsidian-minimal",
    "screenshot": "dark-simple.png",
    "settings": [
        {
            "header": "**Welcome \ud83d\udc4b**: Use the Minimal Theme Settings plugin to access hotkeys, adjust features, select fonts, and choose from preset color schemes. Use the settings below for more granular customization.",
            "type": "heading"
        },
        {
            "header": "**Interface**: ",
            "type": "heading"
        },
        {
            "title": "- Base color: Defines all background and border colors unless overridden below",
            "type": "variable-themed-color"
        },
        {
            "title": "- Accent color: Defines link and checkbox colors unless overridden below",
            "type": "variable-themed-color"
        },
        {
            "title": "- Main window background",
            "type": "variable-themed-color"
        },
        {
            "title": "- Sidebar background",
            "type": "variable-themed-color"
        },
        {
            "title": "- Active background: For hovered buttons and current file",
            "type": "variable-themed-color"
        },
        {
            "title": "- Border color: For divider lines and outlined elements",
            "type": "variable-themed-color"
        },
        {
            "title": "- Border color (highlighted)",
            "type": "variable-themed-color"
        },
        {
            "title": "- Border color (focused)",
            "type": "variable-themed-color"
        },
        {
            "header": "**Icons**: ",
            "type": "heading"
        },
        {
            "title": "- Icon opacity (inactive)",
            "type": "variable-number-slider"
        },
        {
            "title": "- Icon color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Icon color (hover)",
            "type": "variable-themed-color"
        },
        {
            "title": "- Icon color (active)",
            "type": "variable-themed-color"
        },
        {
            "header": "**Links**: ",
            "type": "heading"
        },
        {
            "title": "- Link color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Link color (hovering)",
            "type": "variable-themed-color"
        },
        {
            "title": "- Link font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "header": "**Text**: ",
            "type": "heading"
        },
        {
            "title": "- Normal text color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Selected text background",
            "type": "variable-themed-color"
        },
        {
            "title": "- Muted text color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Faint text color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Bold text color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Italic text color",
            "type": "variable-themed-color"
        },
        {
            "header": "**Headings**: ",
            "type": "heading"
        },
        {
            "title": "- **Level 1 Headings**: "
        },
        {
            "title": "    - H1 font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "    - H1 font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "title": "    - H1 font color",
            "type": "variable-themed-color"
        },
        {
            "title": "    - H1 font variant",
            "type": "variable-select"
        },
        {
            "title": "    - H1 font style",
            "type": "variable-select"
        },
        {
            "title": "- **Level 2 Headings**: "
        },
        {
            "title": "    - H2 font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "    - H2 font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "title": "    - H2 font color",
            "type": "variable-themed-color"
        },
        {
            "title": "    - H2 font variant",
            "type": "variable-select"
        },
        {
            "title": "    - H2 font style",
            "type": "variable-select"
        },
        {
            "title": "- **Level 3 Headings**: "
        },
        {
            "title": "    - H3 font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "    - H3 font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "title": "    - H3 font color",
            "type": "variable-themed-color"
        },
        {
            "title": "    - H3 font variant",
            "type": "variable-select"
        },
        {
            "title": "    - H3 font style",
            "type": "variable-select"
        },
        {
            "title": "- **Level 4 Headings**: "
        },
        {
            "title": "    - H4 font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "    - H4 font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "title": "    - H4 font color",
            "type": "variable-themed-color"
        },
        {
            "title": "    - H4 font variant",
            "type": "variable-select"
        },
        {
            "title": "    - H4 font style",
            "type": "variable-select"
        },
        {
            "title": "- **Level 5 Headings**: "
        },
        {
            "title": "    - H5 font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "    - H5 font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "title": "    - H5 font color",
            "type": "variable-themed-color"
        },
        {
            "title": "    - H5 font variant",
            "type": "variable-select"
        },
        {
            "title": "    - H5 font style",
            "type": "variable-select"
        },
        {
            "title": "- **Level 6 Headings**: "
        },
        {
            "title": "    - H6 font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "    - H6 font weight: Accepts numbers representing the CSS font-weight",
            "type": "variable-number"
        },
        {
            "title": "    - H6 font color",
            "type": "variable-themed-color"
        },
        {
            "title": "    - H6 font variant",
            "type": "variable-select"
        },
        {
            "title": "    - H6 font style",
            "type": "variable-select"
        },
        {
            "header": "**Blockquotes**: ",
            "type": "heading"
        },
        {
            "title": "- Blockquotes text color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Blockquotes font size: Accepts any CSS font-size value",
            "type": "variable-text"
        },
        {
            "title": "- Blockquotes font style",
            "type": "variable-select"
        },
        {
            "header": "**Lists and checkboxes**: ",
            "type": "heading"
        },
        {
            "title": "- Checkbox color",
            "type": "variable-themed-color"
        },
        {
            "title": "- Strike completed tasks: Use line through greyed text for completed tasks",
            "type": "class-toggle"
        },
        {
            "title": "- List item spacing: Space between list items in em units",
            "type": "variable-number-slider"
        },
        {
            "header": "**Code blocks**: ",
            "type": "heading"
        },
        {
            "title": "- Code text color: Where syntax highlighting is not present",
            "type": "variable-themed-color"
        },
        {
            "title": "- Code font size: Accepts any CSS font-size value",
            "type": "variable-text"
        }
    ],
    "user": "kepano"
}

